[BUG] Failed to expand bicep source, accessed JObject values with invalid key value: Reader. Object property name expected

### Existing rule

_No response_

### Description of the issue

The following code leads to PSRule throwing:

```bicep
@description('The role assignments to the Entra ID groups.')
module modPtnRbacSubscription 'br:<placeholder>.azurecr.io/res/authorization/role-assignment:0.4.1' = [for role in parRbacRoles : {
  name: 'modPtnRbac-${varTargetScope}-${role}-${varUniqueString}'
  params: {
    principalId: modPtnDSEntraIDGroups.outputs.outputs[role]
    roleDefinitionIdOrName: role
    principalType: 'Group'
    managementGroupId: parTargetScopeForRoleAssignments == 'managementGroup' ? parRbacTarget.managementGroupId : ''
    subscriptionId: parTargetScopeForRoleAssignments == 'resourceGroup' || parTargetScopeForRoleAssignments == 'subscription' ? parRbacTarget.subscriptionId : ''
    resourceGroupName: parTargetScopeForRoleAssignments == 'resourceGroup' ? parRbacTarget.resourceGroupName : ''
  }
}]
```

### Error messages

```
2024-03-12T10:42:05.3644495Z ##[error]Failed to expand bicep source '/home/runner/work/platform/platform/./deployments/sub/pro-env-001/parameters/main.dev.bicepparam'. Exception calling "GetBicepParamResources" with "2" argument(s): "Unable to expand resources because the source file '/home/runner/work/platform/platform/./deployments/sub/pro-env-001/parameters/main.dev.bicepparam' was not valid. An error occurred evaluating expression '[reference('modPtnDSEntraIDGroups').outputs.outputs.value[parameters('parRbacRoles')[copyIndex()]]]' line 15984. Accessed JObject values with invalid key value: Reader. Object property name expected."
2024-03-12T10:42:05.4808144Z 
2024-03-12T10:42:05.4809392Z Rules processed: 0, failed: 0, errored: 0
2024-03-12T10:42:05.4811076Z Run organization/platform/8247435375 completed in 00:00:05.6050227
2024-03-12T10:42:05.5448321Z Output written to the following file: '/home/runner/work/platform/platform/reports/ps-rule-results.sarif'
2024-03-12T10:42:05.5511541Z ##[error]One or more assertions failed. One or more errors occured.
```

### Reproduction

```bicep
@description('The deployment script to create the Entra ID groups.')
module modPtnDSEntraIDGroups 'br/public:avm/res/resources/deployment-script:0.1.2' = {
  scope: resourceGroup(varDeploymentScriptLocation.subscriptionId, varDeploymentScriptLocation.resourceGroupName)
  name: 'modPtnDSEntraIDGroups-${varUniqueString}'
  params: {
    kind: 'AzurePowerShell'
    name: toLower('ds-${varUniqueString}-${varEnvironment}-${varRegionCode}-001')
    managedIdentities: {
      userAssignedResourcesIds: [
        resPtnExistingUserAssignedIdentity.id
      ]
    }
    tags: {
      'hidden-title': 'Create Entra ID groups for ${varTargetScope} ${varTargetName}'
    }
    retentionInterval: 'PT1H'
    azPowerShellVersion: '9.7'
    arguments: '-Scope ${varTargetScope} -Name ${varTargetName} -Roles ${join(parRbacRoles, ',')} -SubscriptionName ${parRbacTarget.subscriptionName}'
    scriptContent: '''
      [CmdletBinding()]
      param (
        [Parameter()]
        $Name,
        $Scope,
        $Roles,
        $SubscriptionName
      )
      
      $DeploymentScriptOutputs = @{}
      
      foreach ($role in $Roles) {
      
        $groupParams = @{
          DisplayName     = if ($Scope -eq 'rg') { "AZ RBAC $Scope $Name $role ($SubscriptionName)" } else { "AZ RBAC $Scope $Name $role" }
          Description     = if ($Scope -eq 'rg') { "Grant $role access to $Scope $Name ($SubscriptionName)" } else { "Grant $role access to $Scope $Name" }
          MailEnabled     = $false
          SecurityEnabled = $true
          MailNickName    = if ($Scope -eq 'rg') { ("azrbac$Scope$Name$role$SubscriptionName" -replace '-').tolower() } else { ("azrbac$Scope$Name$role" -replace '-').tolower() }
        }
      
        $exists = Get-AzADGroup -DisplayName $groupParams.DisplayName
      
        if ($exists) {
          Write-Output "~ Updating group $($exists.DisplayName)"
          $null = Update-AzADGroup -ObjectId $exists.id @groupParams
        }
        else {
          Write-Output "+ Creating group $($groupParams.DisplayName)"
          $null = New-AzADGroup @groupParams
        }
      
        do {
          try {
            $group = Get-AzADGroup -DisplayName $groupParams.DisplayName
            $groupFound = $true
            # Output the group ids to be used in the bicep file
            $DeploymentScriptOutputs[$role] = $group.id
          }
          catch {
            Write-Host -Object "$($groupParams.DisplayName) not found, will retry in 5 seconds"
            start-sleep -Seconds 5
          }
        } While (!$group)
      }
      
    '''
  }
}

@description('The role assignments to the Entra ID groups.')
module modPtnRbacSubscription 'br:<placeholder>.azurecr.io/res/authorization/role-assignment:0.4.1' = [for role in parRbacRoles : {
  name: 'modPtnRbac-${varTargetScope}-${role}-${varUniqueString}'
  params: {
    principalId: modPtnDSEntraIDGroups.outputs.outputs[role]
    roleDefinitionIdOrName: role
    principalType: 'Group'
    managementGroupId: parTargetScopeForRoleAssignments == 'managementGroup' ? parRbacTarget.managementGroupId : ''
    subscriptionId: parTargetScopeForRoleAssignments == 'resourceGroup' || parTargetScopeForRoleAssignments == 'subscription' ? parRbacTarget.subscriptionId : ''
    resourceGroupName: parTargetScopeForRoleAssignments == 'resourceGroup' ? parRbacTarget.resourceGroupName : ''
  }
}]
```

### Version of PSRule

2.9.0

### Version of PSRule for Azure

1.34.2

### Additional context

_No response_

--- a/docs/CHANGELOG-v1.md
+++ b/docs/CHANGELOG-v1.md
@@ -37,6 +37,9 @@ What's changed since pre-release v1.35.0-B0030:
 - General improvements:
   - Quality updates to rule documentation by @BernieWhite.
     [#2570](https://github.com/Azure/PSRule.Rules.Azure/issues/2570)
+- Bug fixes:
+  - Fixed failed to expand JObject value with invalid key by @BernieWhite.
+    [#2751](https://github.com/Azure/PSRule.Rules.Azure/issues/2751)
 
 ## v1.35.0-B0030 (pre-release)
 

--- a/src/PSRule.Rules.Azure/Data/Template/Mocks.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/Mocks.cs
@@ -318,6 +318,7 @@ namespace PSRule.Rules.Azure.Data.Template
 
             public virtual JToken GetValue(object key)
             {
+                key = GetBaseObject(key);
                 var result = base[key];
                 if (result == null)
                 {
@@ -510,6 +511,20 @@ namespace PSRule.Rules.Azure.Data.Template
                 return true;
             }
             return false;
+        }
+
+        /// <summary>
+        /// Unwrap the base object from a JToken.
+        /// </summary>
+        private static object GetBaseObject(object o)
+        {
+            if (o is JToken t_string && t_string.Type == JTokenType.String)
+                return t_string.Value<string>();
+
+            if (o is JToken t_int && t_int.Type == JTokenType.Integer)
+                return t_int.Value<int>();
+
+            return o;
         }
     }
 }

--- a/src/PSRule.Rules.Azure/Data/Template/TemplateVisitor.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/TemplateVisitor.cs
@@ -11,6 +11,7 @@ using Newtonsoft.Json.Linq;
 using PSRule.Rules.Azure.Configuration;
 using PSRule.Rules.Azure.Pipeline;
 using PSRule.Rules.Azure.Resources;
+using static PSRule.Rules.Azure.Data.Template.Mock;
 
 namespace PSRule.Rules.Azure.Data.Template
 {
@@ -910,7 +911,7 @@ namespace PSRule.Rules.Azure.Data.Template
             public object GetValue()
             {
                 ResolveProperty(_Context, _Value, PROPERTY_VALUE);
-                return _Value;
+                return new MockObject(_Value);
             }
         }
 

--- a/tests/PSRule.Rules.Azure.Tests/PSRule.Rules.Azure.Tests.csproj
+++ b/tests/PSRule.Rules.Azure.Tests/PSRule.Rules.Azure.Tests.csproj
@@ -236,6 +236,9 @@
     <None Update="Tests.Bicep.34.json">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Update="Tests.Bicep.35.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
     <None Update="Tests.Bicep.4.json">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>

--- a/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTests.cs
@@ -13,6 +13,9 @@ using static PSRule.Rules.Azure.Data.Template.TemplateVisitor;
 
 namespace PSRule.Rules.Azure
 {
+    /// <summary>
+    /// This class tests the functionality of the <see cref="TemplateVisitor"/> class that is used to expand resources from an ARM template/ Bicep file.
+    /// </summary>
     public sealed class TemplateVisitorTests
     {
         [Fact]
@@ -1090,6 +1093,15 @@ namespace PSRule.Rules.Azure
 
             actual = resources[3];
             Assert.Equal("Microsoft.KeyVault/vaults/accessPolicies", actual["type"].Value<string>());
+        }
+
+        /// <summary>
+        /// Test case for https://github.com/Azure/PSRule.Rules.Azure/issues/2751.
+        /// </summary>
+        [Fact]
+        public void Index_into_mock_output_object()
+        {
+            var resources = ProcessTemplate(GetSourcePath("Tests.Bicep.35.json"), null, out _);
         }
 
         #region Helper methods

--- a/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.35.bicep
+++ b/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.35.bicep
@@ -1,0 +1,31 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Test case for https://github.com/Azure/PSRule.Rules.Azure/issues/2751.
+
+var roles = ['Reader']
+
+module script 'br/public:avm/res/resources/deployment-script:0.1.2' = {
+  name: 'script'
+  params: {
+    kind: 'AzurePowerShell'
+    name: toLower('ds-001')
+    retentionInterval: 'PT1H'
+    azPowerShellVersion: '9.7'
+    arguments: ''
+    scriptContent: '''
+      
+    '''
+  }
+}
+
+module other './Tests.Bicep.35.child.bicep' = [
+  for role in roles: {
+    name: role
+    params: {
+      value: script.outputs.outputs[role]
+    }
+  }
+]
+
+output other string = other[0].outputs.value

--- a/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.35.child.bicep
+++ b/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.35.child.bicep
@@ -1,0 +1,6 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+param value string
+
+output value string = value

--- a/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.35.json
+++ b/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.35.json
@@ -1,0 +1,551 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "metadata": {
+    "_generator": {
+      "name": "bicep",
+      "version": "0.26.54.24096",
+      "templateHash": "8940134219554320480"
+    }
+  },
+  "variables": {
+    "roles": [
+      "Reader"
+    ]
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2022-09-01",
+      "name": "script",
+      "properties": {
+        "expressionEvaluationOptions": {
+          "scope": "inner"
+        },
+        "mode": "Incremental",
+        "parameters": {
+          "kind": {
+            "value": "AzurePowerShell"
+          },
+          "name": {
+            "value": "[toLower('ds-001')]"
+          },
+          "retentionInterval": {
+            "value": "PT1H"
+          },
+          "azPowerShellVersion": {
+            "value": "9.7"
+          },
+          "arguments": {
+            "value": ""
+          },
+          "scriptContent": {
+            "value": "      \r\n    "
+          }
+        },
+        "template": {
+          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+          "languageVersion": "2.0",
+          "contentVersion": "1.0.0.0",
+          "metadata": {
+            "_generator": {
+              "name": "bicep",
+              "version": "0.23.1.45101",
+              "templateHash": "8734511554398837213"
+            },
+            "name": "Deployment Scripts",
+            "description": "This module deploys Deployment Scripts.",
+            "owner": "Azure/module-maintainers"
+          },
+          "definitions": {
+            "lockType": {
+              "type": "object",
+              "properties": {
+                "name": {
+                  "type": "string",
+                  "nullable": true,
+                  "metadata": {
+                    "description": "Optional. Specify the name of lock."
+                  }
+                },
+                "kind": {
+                  "type": "string",
+                  "allowedValues": [
+                    "CanNotDelete",
+                    "None",
+                    "ReadOnly"
+                  ],
+                  "nullable": true,
+                  "metadata": {
+                    "description": "Optional. Specify the type of lock."
+                  }
+                }
+              },
+              "nullable": true
+            },
+            "managedIdentitiesType": {
+              "type": "object",
+              "properties": {
+                "userAssignedResourcesIds": {
+                  "type": "array",
+                  "items": {
+                    "type": "string"
+                  },
+                  "metadata": {
+                    "description": "Optional. The resource ID(s) to assign to the resource."
+                  }
+                }
+              },
+              "nullable": true
+            },
+            "roleAssignmentType": {
+              "type": "array",
+              "items": {
+                "type": "object",
+                "properties": {
+                  "roleDefinitionIdOrName": {
+                    "type": "string",
+                    "metadata": {
+                      "description": "Required. The role to assign. You can provide either the display name of the role definition, the role definition GUID, or its fully qualified ID in the following format: '/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11'."
+                    }
+                  },
+                  "principalId": {
+                    "type": "string",
+                    "metadata": {
+                      "description": "Required. The principal ID of the principal (user/group/identity) to assign the role to."
+                    }
+                  },
+                  "principalType": {
+                    "type": "string",
+                    "allowedValues": [
+                      "Device",
+                      "ForeignGroup",
+                      "Group",
+                      "ServicePrincipal",
+                      "User"
+                    ],
+                    "nullable": true,
+                    "metadata": {
+                      "description": "Optional. The principal type of the assigned principal ID."
+                    }
+                  },
+                  "description": {
+                    "type": "string",
+                    "nullable": true,
+                    "metadata": {
+                      "description": "Optional. The description of the role assignment."
+                    }
+                  },
+                  "condition": {
+                    "type": "string",
+                    "nullable": true,
+                    "metadata": {
+                      "description": "Optional. The conditions on the role assignment. This limits the resources it can be assigned to. e.g.: @Resource[Microsoft.Storage/storageAccounts/blobServices/containers:ContainerName] StringEqualsIgnoreCase \"foo_storage_container\"."
+                    }
+                  },
+                  "conditionVersion": {
+                    "type": "string",
+                    "allowedValues": [
+                      "2.0"
+                    ],
+                    "nullable": true,
+                    "metadata": {
+                      "description": "Optional. Version of the condition."
+                    }
+                  },
+                  "delegatedManagedIdentityResourceId": {
+                    "type": "string",
+                    "nullable": true,
+                    "metadata": {
+                      "description": "Optional. The Resource Id of the delegated managed identity resource."
+                    }
+                  }
+                }
+              },
+              "nullable": true
+            }
+          },
+          "parameters": {
+            "name": {
+              "type": "string",
+              "maxLength": 24,
+              "metadata": {
+                "description": "Required. Name of the Deployment Script."
+              }
+            },
+            "location": {
+              "type": "string",
+              "defaultValue": "[resourceGroup().location]",
+              "metadata": {
+                "description": "Optional. Location for all resources."
+              }
+            },
+            "kind": {
+              "type": "string",
+              "allowedValues": [
+                "AzureCLI",
+                "AzurePowerShell"
+              ],
+              "metadata": {
+                "description": "Required. Specifies the Kind of the Deployment Script."
+              }
+            },
+            "managedIdentities": {
+              "$ref": "#/definitions/managedIdentitiesType",
+              "metadata": {
+                "description": "Optional. The managed identity definition for this resource."
+              }
+            },
+            "tags": {
+              "type": "object",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Resource tags."
+              }
+            },
+            "azPowerShellVersion": {
+              "type": "string",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Azure PowerShell module version to be used. See a list of supported Azure PowerShell versions: https://mcr.microsoft.com/v2/azuredeploymentscripts-powershell/tags/list."
+              }
+            },
+            "azCliVersion": {
+              "type": "string",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Azure CLI module version to be used. See a list of supported Azure CLI versions: https://mcr.microsoft.com/v2/azure-cli/tags/list."
+              }
+            },
+            "scriptContent": {
+              "type": "string",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Script body. Max length: 32000 characters. To run an external script, use primaryScriptURI instead."
+              }
+            },
+            "primaryScriptUri": {
+              "type": "string",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Uri for the external script. This is the entry point for the external script. To run an internal script, use the scriptContent parameter instead."
+              }
+            },
+            "environmentVariables": {
+              "type": "secureObject",
+              "defaultValue": {},
+              "metadata": {
+                "example": "secureList: [\n  {\n    name: 'string'\n    secureValue: 'string'\n    value: 'string'\n  }\n]\n",
+                "description": "Optional. The environment variables to pass over to the script. The list is passed as an object with a key name \"secureList\" and the value is the list of environment variables (array). The list must have a 'name' and a 'value' or a 'secretValue' property for each object."
+              }
+            },
+            "supportingScriptUris": {
+              "type": "array",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. List of supporting files for the external script (defined in primaryScriptUri). Does not work with internal scripts (code defined in scriptContent)."
+              }
+            },
+            "subnetResourceIds": {
+              "type": "array",
+              "items": {
+                "type": "string"
+              },
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. List of subnet IDs to use for the container group. This is required if you want to run the deployment script in a private network. When using a private network, the `Storage File Data Privileged Contributor` role needs to be assigned to the user-assigned managed identity and the deployment principal needs to have permissions to list the storage account keys. Also, Shared-Keys must not be disabled on the used storage account [ref](https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/deployment-script-vnet)."
+              }
+            },
+            "arguments": {
+              "type": "string",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Command-line arguments to pass to the script. Arguments are separated by spaces."
+              }
+            },
+            "retentionInterval": {
+              "type": "string",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Interval for which the service retains the script resource after it reaches a terminal state. Resource will be deleted when this duration expires. Duration is based on ISO 8601 pattern (for example P7D means one week)."
+              }
+            },
+            "baseTime": {
+              "type": "string",
+              "defaultValue": "[utcNow('yyyy-MM-dd-HH-mm-ss')]",
+              "metadata": {
+                "description": "Generated. Do not provide a value! This date value is used to make sure the script run every time the template is deployed."
+              }
+            },
+            "runOnce": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "Optional. When set to false, script will run every time the template is deployed. When set to true, the script will only run once."
+              }
+            },
+            "cleanupPreference": {
+              "type": "string",
+              "defaultValue": "Always",
+              "allowedValues": [
+                "Always",
+                "OnSuccess",
+                "OnExpiration"
+              ],
+              "metadata": {
+                "description": "Optional. The clean up preference when the script execution gets in a terminal state. Specify the preference on when to delete the deployment script resources. The default value is Always, which means the deployment script resources are deleted despite the terminal state (Succeeded, Failed, canceled)."
+              }
+            },
+            "containerGroupName": {
+              "type": "string",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Container group name, if not specified then the name will get auto-generated. Not specifying a 'containerGroupName' indicates the system to generate a unique name which might end up flagging an Azure Policy as non-compliant. Use 'containerGroupName' when you have an Azure Policy that expects a specific naming convention or when you want to fully control the name. 'containerGroupName' property must be between 1 and 63 characters long, must contain only lowercase letters, numbers, and dashes and it cannot start or end with a dash and consecutive dashes are not allowed."
+              }
+            },
+            "storageAccountResourceId": {
+              "type": "string",
+              "defaultValue": "",
+              "metadata": {
+                "description": "Optional. The resource ID of the storage account to use for this deployment script. If none is provided, the deployment script uses a temporary, managed storage account."
+              }
+            },
+            "timeout": {
+              "type": "string",
+              "nullable": true,
+              "metadata": {
+                "description": "Optional. Maximum allowed script execution time specified in ISO 8601 format. Default value is PT1H - 1 hour; 'PT30M' - 30 minutes; 'P5D' - 5 days; 'P1Y' 1 year."
+              }
+            },
+            "lock": {
+              "$ref": "#/definitions/lockType",
+              "metadata": {
+                "description": "Optional. The lock settings of the service."
+              }
+            },
+            "roleAssignments": {
+              "$ref": "#/definitions/roleAssignmentType",
+              "metadata": {
+                "description": "Optional. Array of role assignments to create."
+              }
+            },
+            "enableTelemetry": {
+              "type": "bool",
+              "defaultValue": true,
+              "metadata": {
+                "description": "Optional. Enable/Disable usage telemetry for module."
+              }
+            }
+          },
+          "variables": {
+            "copy": [
+              {
+                "name": "subnetIds",
+                "count": "[length(coalesce(parameters('subnetResourceIds'), createArray()))]",
+                "input": {
+                  "id": "[coalesce(parameters('subnetResourceIds'), createArray())[copyIndex('subnetIds')]]"
+                }
+              }
+            ],
+            "builtInRoleNames": {
+              "Contributor": "[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b24988ac-6180-42a0-ab88-20f7382dd24c')]",
+              "Owner": "[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')]",
+              "Reader": "[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')]",
+              "Role Based Access Control Administrator (Preview)": "[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'f58310d9-a9f6-439a-9e8d-f62e7b41a168')]",
+              "User Access Administrator": "[subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9')]"
+            },
+            "containerSettings": {
+              "containerGroupName": "[parameters('containerGroupName')]",
+              "subnetIds": "[if(not(empty(coalesce(variables('subnetIds'), createArray()))), variables('subnetIds'), null())]"
+            },
+            "formattedUserAssignedIdentities": "[reduce(map(coalesce(tryGet(parameters('managedIdentities'), 'userAssignedResourcesIds'), createArray()), lambda('id', createObject(format('{0}', lambdaVariables('id')), createObject()))), createObject(), lambda('cur', 'next', union(lambdaVariables('cur'), lambdaVariables('next'))))]",
+            "identity": "[if(not(empty(parameters('managedIdentities'))), createObject('type', if(not(empty(coalesce(tryGet(parameters('managedIdentities'), 'userAssignedResourcesIds'), createObject()))), 'UserAssigned', null()), 'userAssignedIdentities', if(not(empty(variables('formattedUserAssignedIdentities'))), variables('formattedUserAssignedIdentities'), null())), null())]"
+          },
+          "resources": {
+            "storageAccount": {
+              "condition": "[not(empty(parameters('storageAccountResourceId')))]",
+              "existing": true,
+              "type": "Microsoft.Storage/storageAccounts",
+              "apiVersion": "2021-04-01",
+              "subscriptionId": "[split(if(not(empty(parameters('storageAccountResourceId'))), parameters('storageAccountResourceId'), '//'), '/')[2]]",
+              "resourceGroup": "[split(if(not(empty(parameters('storageAccountResourceId'))), parameters('storageAccountResourceId'), '////'), '/')[4]]",
+              "name": "[last(split(if(not(empty(parameters('storageAccountResourceId'))), parameters('storageAccountResourceId'), 'dummyAccount'), '/'))]"
+            },
+            "deploymentScript_lock": {
+              "condition": "[and(not(empty(coalesce(parameters('lock'), createObject()))), not(equals(tryGet(parameters('lock'), 'kind'), 'None')))]",
+              "type": "Microsoft.Authorization/locks",
+              "apiVersion": "2020-05-01",
+              "scope": "[format('Microsoft.Resources/deploymentScripts/{0}', parameters('name'))]",
+              "name": "[coalesce(tryGet(parameters('lock'), 'name'), format('lock-{0}', parameters('name')))]",
+              "properties": {
+                "level": "[coalesce(tryGet(parameters('lock'), 'kind'), '')]",
+                "notes": "[if(equals(tryGet(parameters('lock'), 'kind'), 'CanNotDelete'), 'Cannot delete resource or child resources.', 'Cannot delete or modify the resource or child resources.')]"
+              },
+              "dependsOn": [
+                "deploymentScript"
+              ]
+            },
+            "deploymentScript_roleAssignments": {
+              "copy": {
+                "name": "deploymentScript_roleAssignments",
+                "count": "[length(coalesce(parameters('roleAssignments'), createArray()))]"
+              },
+              "type": "Microsoft.Authorization/roleAssignments",
+              "apiVersion": "2022-04-01",
+              "scope": "[format('Microsoft.Resources/deploymentScripts/{0}', parameters('name'))]",
+              "name": "[guid(resourceId('Microsoft.Resources/deploymentScripts', parameters('name')), coalesce(parameters('roleAssignments'), createArray())[copyIndex()].principalId, coalesce(parameters('roleAssignments'), createArray())[copyIndex()].roleDefinitionIdOrName)]",
+              "properties": {
+                "roleDefinitionId": "[if(contains(variables('builtInRoleNames'), coalesce(parameters('roleAssignments'), createArray())[copyIndex()].roleDefinitionIdOrName), variables('builtInRoleNames')[coalesce(parameters('roleAssignments'), createArray())[copyIndex()].roleDefinitionIdOrName], if(contains(coalesce(parameters('roleAssignments'), createArray())[copyIndex()].roleDefinitionIdOrName, '/providers/Microsoft.Authorization/roleDefinitions/'), coalesce(parameters('roleAssignments'), createArray())[copyIndex()].roleDefinitionIdOrName, subscriptionResourceId('Microsoft.Authorization/roleDefinitions', coalesce(parameters('roleAssignments'), createArray())[copyIndex()].roleDefinitionIdOrName)))]",
+                "principalId": "[coalesce(parameters('roleAssignments'), createArray())[copyIndex()].principalId]",
+                "description": "[tryGet(coalesce(parameters('roleAssignments'), createArray())[copyIndex()], 'description')]",
+                "principalType": "[tryGet(coalesce(parameters('roleAssignments'), createArray())[copyIndex()], 'principalType')]",
+                "condition": "[tryGet(coalesce(parameters('roleAssignments'), createArray())[copyIndex()], 'condition')]",
+                "conditionVersion": "[if(not(empty(tryGet(coalesce(parameters('roleAssignments'), createArray())[copyIndex()], 'condition'))), coalesce(tryGet(coalesce(parameters('roleAssignments'), createArray())[copyIndex()], 'conditionVersion'), '2.0'), null())]",
+                "delegatedManagedIdentityResourceId": "[tryGet(coalesce(parameters('roleAssignments'), createArray())[copyIndex()], 'delegatedManagedIdentityResourceId')]"
+              },
+              "dependsOn": [
+                "deploymentScript"
+              ]
+            },
+            "avmTelemetry": {
+              "condition": "[parameters('enableTelemetry')]",
+              "type": "Microsoft.Resources/deployments",
+              "apiVersion": "2023-07-01",
+              "name": "[format('46d3xbcp.res.resources-deploymentscript.{0}.{1}', replace('0.1.2', '.', '-'), substring(uniqueString(deployment().name, parameters('location')), 0, 4))]",
+              "properties": {
+                "mode": "Incremental",
+                "template": {
+                  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+                  "contentVersion": "1.0.0.0",
+                  "resources": [],
+                  "outputs": {
+                    "telemetry": {
+                      "type": "String",
+                      "value": "For more information, see https://aka.ms/avm/TelemetryInfo"
+                    }
+                  }
+                }
+              }
+            },
+            "deploymentScript": {
+              "type": "Microsoft.Resources/deploymentScripts",
+              "apiVersion": "2023-08-01",
+              "name": "[parameters('name')]",
+              "location": "[parameters('location')]",
+              "tags": "[parameters('tags')]",
+              "identity": "[variables('identity')]",
+              "kind": "[parameters('kind')]",
+              "properties": {
+                "azPowerShellVersion": "[if(equals(parameters('kind'), 'AzurePowerShell'), parameters('azPowerShellVersion'), null())]",
+                "azCliVersion": "[if(equals(parameters('kind'), 'AzureCLI'), parameters('azCliVersion'), null())]",
+                "containerSettings": "[if(not(empty(variables('containerSettings'))), variables('containerSettings'), null())]",
+                "storageAccountSettings": "[if(not(empty(parameters('storageAccountResourceId'))), if(not(empty(parameters('storageAccountResourceId'))), createObject('storageAccountKey', if(empty(parameters('subnetResourceIds')), listKeys(extensionResourceId(format('/subscriptions/{0}/resourceGroups/{1}', split(if(not(empty(parameters('storageAccountResourceId'))), parameters('storageAccountResourceId'), '//'), '/')[2], split(if(not(empty(parameters('storageAccountResourceId'))), parameters('storageAccountResourceId'), '////'), '/')[4]), 'Microsoft.Storage/storageAccounts', last(split(if(not(empty(parameters('storageAccountResourceId'))), parameters('storageAccountResourceId'), 'dummyAccount'), '/'))), '2023-01-01').keys[0].value, null()), 'storageAccountName', last(split(parameters('storageAccountResourceId'), '/'))), null()), null())]",
+                "arguments": "[parameters('arguments')]",
+                "environmentVariables": "[if(not(empty(parameters('environmentVariables'))), parameters('environmentVariables').secureList, createArray())]",
+                "scriptContent": "[if(not(empty(parameters('scriptContent'))), parameters('scriptContent'), null())]",
+                "primaryScriptUri": "[if(not(empty(parameters('primaryScriptUri'))), parameters('primaryScriptUri'), null())]",
+                "supportingScriptUris": "[if(not(empty(parameters('supportingScriptUris'))), parameters('supportingScriptUris'), null())]",
+                "cleanupPreference": "[parameters('cleanupPreference')]",
+                "forceUpdateTag": "[if(parameters('runOnce'), resourceGroup().name, parameters('baseTime'))]",
+                "retentionInterval": "[parameters('retentionInterval')]",
+                "timeout": "[parameters('timeout')]"
+              },
+              "dependsOn": [
+                "storageAccount"
+              ]
+            }
+          },
+          "outputs": {
+            "resourceId": {
+              "type": "string",
+              "metadata": {
+                "description": "The resource ID of the deployment script."
+              },
+              "value": "[resourceId('Microsoft.Resources/deploymentScripts', parameters('name'))]"
+            },
+            "resourceGroupName": {
+              "type": "string",
+              "metadata": {
+                "description": "The resource group the deployment script was deployed into."
+              },
+              "value": "[resourceGroup().name]"
+            },
+            "name": {
+              "type": "string",
+              "metadata": {
+                "description": "The name of the deployment script."
+              },
+              "value": "[parameters('name')]"
+            },
+            "location": {
+              "type": "string",
+              "metadata": {
+                "description": "The location the resource was deployed into."
+              },
+              "value": "[reference('deploymentScript', '2023-08-01', 'full').location]"
+            },
+            "outputs": {
+              "type": "object",
+              "metadata": {
+                "description": "The output of the deployment script."
+              },
+              "value": "[if(contains(reference('deploymentScript'), 'outputs'), reference('deploymentScript').outputs, createObject())]"
+            }
+          }
+        }
+      }
+    },
+    {
+      "copy": {
+        "name": "other",
+        "count": "[length(variables('roles'))]"
+      },
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2022-09-01",
+      "name": "[variables('roles')[copyIndex()]]",
+      "properties": {
+        "expressionEvaluationOptions": {
+          "scope": "inner"
+        },
+        "mode": "Incremental",
+        "parameters": {
+          "value": {
+            "value": "[reference(resourceId('Microsoft.Resources/deployments', 'script'), '2022-09-01').outputs.outputs.value[variables('roles')[copyIndex()]]]"
+          }
+        },
+        "template": {
+          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+          "contentVersion": "1.0.0.0",
+          "metadata": {
+            "_generator": {
+              "name": "bicep",
+              "version": "0.26.54.24096",
+              "templateHash": "4801530143957603952"
+            }
+          },
+          "parameters": {
+            "value": {
+              "type": "string"
+            }
+          },
+          "resources": [],
+          "outputs": {
+            "value": {
+              "type": "string",
+              "value": "[parameters('value')]"
+            }
+          }
+        }
+      },
+      "dependsOn": [
+        "[resourceId('Microsoft.Resources/deployments', 'script')]"
+      ]
+    }
+  ],
+  "outputs": {
+    "other": {
+      "type": "string",
+      "value": "[reference(resourceId('Microsoft.Resources/deployments', variables('roles')[0]), '2022-09-01').outputs.value.value]"
+    }
+  }
+}

